### PR TITLE
Add vscode API `DebugSessionOptions`

### LINKED ISSUE #8518: vscode API completeness: startDebugging to respect DebugSessionOptions param
### Bug Description:

vscode API `startDebugging` has optional parameter of type `DebugSessionOptions` which currently theia ignore:

https://github.com/microsoft/vscode/blob/master/src/vs/vscode.d.ts#L11512

It has several options like no debug, console mode, etc.

### Steps to Reproduce:

Call `startDebugging` from an extension and make sure all options in `DebugSessionOptions` are treated. Example for such options: https://github.com/microsoft/vscode-js-debug/blob/088e5a28b2f97205fc5fca1f19b4cd7657685d7d/src/ui/vsCodeSessionManager.ts#L38-L43

### Additional Information

- Operating System: all
- Theia Version: master

### PR DESCRIPTION
**Pull request by msujew**
#### What it does
Closes #8518 

Adds the `vscode.debug.DebugSessionOptions` API to Theia. This enables the following features:

- Show child debug sessions as a tree under the parent sessions (Theia used to show them as siblings)
    - The new `compact` property will simplify the label and tree structure if possible
- Different debug sessions no longer share the same `DebugConsoleSession` and therefore are displayed independently
    - Child debug sessions with the `consoleMode` property set to `MergeWithParent` will continue to share their console session with each other.
    - You can switch between console sessions with a selector, assuming more than one session is currently running.

The changes to `DebugConsoleSession` are slightly breaking. While it's still possible to simply inject the `DebugConsoleSession` directly, its tight coupling to `DebugSession` makes that impractical. Instead of injecting the `DebugConsoleSession` directly, a `DebugConsoleSessionFactory` allows to create a `DebugConsoleSession` with a provided `DebugSession`.

#### How to test

1. Clone [this test repo](https://github.com/msujew/debug-session-options-example) into Theia and make sure the `vscode-js-debug` extension is installed in Theia.
2. Start a debugging session with the contained `Launch Program` configuration.
3. Child processes will now be shown as a tree in the threads view.
4. Start a second debugging session with the same configuration
5. Another tree will appear in the threads view and you can switch between the different debug consoles.
6. You can add `configuration.consoleMode = undefined;` and/or `configuration.compact = undefined;` after [this line](https://github.com/msujew/theia/blob/e26d3bff923e51f013aece29ed7a9db8eb27efce/packages/plugin-ext/src/main/browser/debug/debug-main.ts#L273), and restart the server to test how these properties influence the debug console/view.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/debug/src/browser/console/debug-console-session.ts
Comment (colin-grant-work) [inline packages/debug/src/browser/console/debug-console-session.ts:26]: Why have you made this class no longer `injectable`? To facilitate extensibility, we generally want to add `injectability` rather than removing t?

Comment (msujew) [reply packages/debug/src/browser/console/debug-console-session.ts]: Right, didn't really think about that. I made `DebugConsoleSession` injectable again and added a `DebugConsoleSessionProvider` that directly sets the internal `debugSession` to the provided value.

**Comment by marcdumais-work** [comment]
@paul-marechal @vince-fugnitto  Could you help review?

File: packages/console/src/browser/console-session-manager.ts
Comment (vince-fugnitto) [inline packages/console/src/browser/console-session-manager.ts:36]: Given the [coding guidelines](https://github.com/eclipse-theia/theia/wiki/Coding-Guidelines#event_names) we should follow the following format: `on[Will|Did]VerbNoun`.

```suggestion
    readonly onDidAddSession = this.sessionAddedEmitter.event;
```

The same is true for other event names.

**Review by vince-fugnitto — COMMENTED** [review]
The changes look very promising, I only had the chance to quickly review so far and both the code and behavior look correct. I'd like to take some additional time looking more deeply at the code and reviewing additional use-cases.

File: packages/console/src/browser/console-session-manager.ts
Comment (msujew) [reply packages/console/src/browser/console-session-manager.ts]: Thanks! I actually didn't know about that. Done.

File: packages/debug/src/browser/view/debug-threads-source.tsx
Comment (paul-marechal) [inline packages/debug/src/browser/view/debug-threads-source.tsx:48]: Since we're using iterators, why not do:

```ts
    *getElements(): IterableIterator<TreeElement> {
        if (this.model.sessionCount === 1 && this.model.session && this.model.session.threadCount) {
            return yield* this.model.session.threads;
        }
        for (const session of this.model.sessions) {
            if (session.parentSession) {
                yield session;
            }
        }
    }
```

`Array.values` seems to work but I was surprised to see it.

File: packages/debug/src/browser/debug-session.tsx
Comment (msujew) [reply packages/debug/src/browser/debug-session.tsx:72]: I added the breaking change to the changelog. I didn't want to split the constructor definition from `(<properties>, <services>` to `(<properties>, <services>, <parent>)`. Does that make sense, or would you rather see a non-breaking change at that point?

File: packages/console/src/browser/console-session-manager.ts
Comment (msujew) [reply packages/console/src/browser/console-session-manager.ts]: Sure, I can see that. They're all getters now.

File: packages/debug/src/browser/view/debug-threads-source.tsx
Comment (msujew) [reply packages/debug/src/browser/view/debug-threads-source.tsx]: Converted this method and another one in `debug-session.tsx` to use iterators.

File: packages/plugin-ext/compile.tsconfig.json
Comment (vince-fugnitto) [inline packages/plugin-ext/compile.tsconfig.json:78]: minor: do you mind updating the location so it is in alphabetical order like the rest?

**Review by vince-fugnitto — COMMENTED** [review]
I confirmed that the behavior works as expected:
- confirmed with the test workspace that launching multiple debug sessions works correctly, the threads display nested processes, and the output view contributes a toolbar item to select the different sessions.
- confirmed that normal debug sessions (ex: _'run mocha tests') works as before

File: packages/plugin-ext/src/main/browser/debug/debug-main.ts
Comment (msujew) [inline packages/plugin-ext/src/main/browser/debug/debug-main.ts:122]: A quick note on that: I'm wondering whether this makes sense at all. Do we somehow know which extension a debug console belongs to? Currently, it will just log any input to the currently selected console session, but I'm pretty sure that's not the expected behavior. @vince-fugnitto WDYT?

Comment (vince-fugnitto) [reply packages/plugin-ext/src/main/browser/debug/debug-main.ts:122]: I'm not sure if we have the information or can deduce it. I believe on `master` before your changes all output was displayed in the same console. Your changes are more aligned with the behavior I notice in vscode.

Comment (msujew) [reply packages/plugin-ext/src/main/browser/debug/debug-main.ts:122]: Ha, I wanted to see how vscode does it. I think I accidentally replicated [their code](
https://github.com/microsoft/vscode/blob/e657d42281487040b50374d482d34b3fb05eb377/src/vs/workbench/api/browser/mainThreadDebugService.ts#L281-L287). It seems like vscode also just logs to the selected session if `append` or `appendLine` calls are made. So I guess that resolves that question.

I also couldn't find any vscode extension that actually calls these methods.

**Comment by vince-fugnitto** [comment]
@msujew I believe if rebased it should fix the CI failures which have been resolved :+1:

**Review by vince-fugnitto — APPROVED** [review]
I confirmed last week that the changes work as intended and the behavior is aligned with vscode :+1: The code is also very clean and I have no comments.

I'll let others review if interested :)

**Comment by vince-fugnitto** [comment]
@paul-marechal @colin-grant-work any additional comments? Else I'll merge the pull-request early this week.

**Comment by vince-fugnitto** [comment]
@msujew the pull-request just needs to be rebased due to https://github.com/eclipse-theia/theia/pull/8706.

**Comment by msujew** [comment]
@vince-fugnitto Thanks for the heads-up. Rebasing done 👍

**Comment by vince-fugnitto** [comment]
I'll merge the pull-request tomorrow if there are no objections 👍

**Pull request merged**
